### ISSUE TITLE
rgb curve module crashes darktable on second curve controll point

### ISSUE THREAD

**Issue by robMate** [darktable-org/darktable#15358]
### Describe the bug

I recently switched from an Nvidia GPU to an AMD one. Now the rgb curve module causes darktable to crash after i add a second curve control point to the curve.

The crash is reproducible on Wayland & Xorg and with and without opencl.  

![Bildschirmfoto vom 2023-10-01 17-14-08](https://github.com/darktable-org/darktable/assets/5657121/9953d539-8578-4c62-b441-693048ab5c87)

opencl package: opencl-amd 5.7 (rocm-5.7.0)
OS: Manjaro
dt build: 4.5.0+786~g1439bb0eb6
Image:  (could not upload a example file but the issue appeared on RAF & ARW files. I think its not dependent on the file type)
darktable debug output: [darktable_debug_all.txt](https://github.com/darktable-org/darktable/files/12778159/darktable_debug_all.txt)

inxi -Gazy
```
inxi -Gazy
Graphics:
  Device-1: AMD Navi 31 [Radeon RX 7900 XT/7900 XTX] vendor: Sapphire PULSE
    driver: amdgpu v: kernel arch: RDNA-3 code: Navi-3x process: TSMC n5 (5nm)
    built: 2022+ pcie: gen: 4 speed: 16 GT/s lanes: 16 ports: active: DP-1
    empty: DP-2,HDMI-A-1,HDMI-A-2 bus-ID: 03:00.0 chip-ID: 1002:744c
    class-ID: 0300
  Device-2: AMD Raphael driver: amdgpu v: kernel arch: RDNA-2 code: Navi-2x
    process: TSMC n7 (7nm) built: 2020-22 pcie: gen: 4 speed: 16 GT/s lanes: 16
    ports: active: none empty: DP-3, DP-4, DP-5, HDMI-A-3 bus-ID: 0f:00.0
    chip-ID: 1002:164e class-ID: 0300 temp: 60.0 C
  Display: wayland server: X.org v: 1.21.1.8 with: Xwayland v: 23.2.0
    compositor: gnome-shell v: 44.5 driver: gpu: amdgpu display-ID: 0
  Monitor-1: DP-1 model: Samsung C27HG7x serial: <filter> built: 2038
    res: 2560x1440 dpi: 109 gamma: 1.2 size: 598x336mm (23.54x13.23")
    diag: 686mm (27") ratio: 16:9 modes: max: 2560x1440 min: 720x400
  API: EGL v: 1.5 hw: drv: amd radeonsi platforms: device: 0 drv: radeonsi
    device: 1 drv: radeonsi device: 2 drv: swrast gbm: drv: kms_swrast
    surfaceless: drv: radeonsi wayland: drv: radeonsi x11: drv: radeonsi
  API: OpenGL v: 4.6 compat-v: 4.5 vendor: amd mesa v: 23.1.7-1 glx-v: 1.4
    direct-render: yes renderer: AMD Radeon RX 7900 XTX (gfx1100 LLVM 16.0.6 DRM
    3.54 6.5.3-1-MANJARO) device-ID: 1002:744c memory: 23.44 GiB unified: no
    display-ID: :0.0
```

**gdb darktable ( full log [dartable_gdb_output.txt](https://github.com/darktable-org/darktable/files/12778114/dartable_gdb_output.txt) ):**
```
....
[New Thread 0x7ffd9ff8f6c0 (LWP 66627)]
[New Thread 0x7ffd9ef8d6c0 (LWP 66628)]
[Thread 0x7ffd9ff8f6c0 (LWP 66627) exited]
[Thread 0x7ffd9f78e6c0 (LWP 66626) exited]
[Thread 0x7ffd9ef8d6c0 (LWP 66628) exited]

Thread 11 "worker res 0" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffc77fe6c0 (LWP 66292)]
0x00007fffc471f354 in _generate_curve_lut (pipe=<optimized out>, d=d@entry=0x7ffe5113f010)
    at /home/robin/git/darktable/src/iop/rgbcurve.c:1565
1565	        (void)dt_draw_curve_add_point(d->curve[ch], curve_nodes[ch][k].x, curve_nodes[ch][k].y);
(gdb) s

Thread 12 "worker res 1" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffc6ffd6c0 (LWP 66293)]
0x00007fffc471f354 in _generate_curve_lut (pipe=<optimized out>, d=d@entry=0x7ffe3c015010)
    at /home/robin/git/darktable/src/iop/rgbcurve.c:1565
1565	        (void)dt_draw_curve_add_point(d->curve[ch], curve_nodes[ch][k].x, curve_nodes[ch][k].y);
(gdb) 
Couldn't get registers: Kein passender Prozess gefunden.
(gdb) 
Couldn't get registers: Kein passender Prozess gefunden.
(gdb) 
Couldn't get registers: Kein passender Prozess gefunden.
(gdb) 
Couldn't get registers: Kein passender Prozess gefunden.
(gdb) n
Couldn't get registers: Kein passender Prozess gefunden.
(gdb) [Thread 0x7ffdad2496c0 (LWP 66618) exited]
[Thread 0x7ffdada4a6c0 (LWP 66617) exited]
[Thread 0x7ffdae24b6c0 (LWP 66616) exited]
```

Let me know if i can help with additional information. 

### Steps to reproduce

1. open darktable
2. import image
3. open image in darkroom
4. add module "rgb curve"
5. add curve point
6. add second curve point 
7. ...crash 

### Expected behavior

should not crash 

### Where did you install darktable from?

self compiled

### darktable version

4.5.0+786~g1439bb0eb6

### What OS are you using?

Linux

### What is the version of your OS?

Manjaro

### Describe your system?

RAM: 32 GB
CPU: AMD Ryzen™ 9 7900X3D, 12 Core, 24 Threads
GPU: AMD Radeon™ RX 7900 XTX
Display-Server: Wayland
GTK: 2.24.38-1

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

 AMD Radeon™ RX 7900 XTX / amdgpu

**Comment by gi-man** [comment]
Files and original conversation here: 

https://discuss.pixls.us/t/rgb-curve-module-crashes-darktable-on-second-curve-controll-point/39729

**Comment by TurboGit** [comment]
You said since you switched from NVidia to AMD, but this also reproduce without OpenCL. So it rules out an OpenCL issue. Yet, I cannot reproduce and I'm sure it works fine for many others as this is quite a common action (adding a node in RGB Curve)... So there is probably an issue coming from something else, we're missing part of the context here I think.

**Comment by TurboGit** [comment]
Even with your RAW + xmp I cannot reproduce the crash.

**Comment by robMate** [comment]
hm its the same for me. I tested it on my laptop and everything works. I compared the build logs of both devices but could not find a meaningful difference like a missing library. Both devices run on manjaro. Currently, I'm looking in gdb debugging to see if i can provide some more information.

**Comment by robMate** [comment]
I found out that it dose not crash if i compile as a debug build like this 
```
./build.sh --prefix /opt/darktable --build-type Debug --install --sudo
```

I tried a Release build after this, but it crashed on the same line as usual?
```
...
[New Thread 0x7ffde7aa46c0 (LWP 56082)]
[New Thread 0x7ffde72a36c0 (LWP 56083)]

Thread 6 "worker 0" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffdeffd6c0 (LWP 55948)]
0x00007fffbc52f354 in _generate_curve_lut (pipe=<optimized out>, d=d@entry=0x7ffea4f3f010) at /home/robin/git/darktable/src/iop/rgbcurve.c:1565
1565	        (void)dt_draw_curve_add_point(d->curve[ch], curve_nodes[ch][k].x, curve_nodes[ch][k].y);
(gdb) info locals
k = <optimized out>
ch = 0
work_profile = <optimized out>
curve_nodes = {{{x = 0, y = 0}, {x = 0.105990782, y = 0.10599079}, {x = 0.235023037, y = 0.235023037}, {x = 0.374172181, y = 0.374172181}, {x = 1, y = 1}, {x = 0, y = 0} <repeats 15 times>}, {{x = 0, y = 0}, {x = 1, y = 1}, {x = 0, y = 0} <repeats 18 times>}, {{x = 0, y = 0}, {
      x = 1, y = 1}, {x = 0, y = 0} <repeats 18 times>}}
(gdb) 
``

**Comment by gi-man** [comment]
I also cannot reproduce using your image and xmp. Fedora 38 KDE current master.

What gcc version?

**Comment by robMate** [comment]
Here is also a backtrace file of the crash: [darktable_bt_8832B2.txt](https://github.com/darktable-org/darktable/files/12785608/darktable_bt_8832B2.txt)

I not shure why it works on a debug build but not on a release build. Has it something to do with the optimization for the AMD CPU? ( Wild guess, have no idea what happens there)

**Comment by gi-man** [comment]
I have an AMD CPU too. 

Can you try to do the same edits but with the `Tone Curve` module instead? It uses a similar interface for the points.

**Comment by robMate** [comment]
I tested it, but it works there without any issues. I also tested the base curve, works.
On the rgb curve i also edited the channels separate and as soon as i add two points on one of the channels darktable crashes.

**Comment by robMate** [comment]
I added a screencast of the issue on discuss.pixls.us because i could not upload it here https://discuss.pixls.us/t/rgb-curve-module-crashes-darktable-on-second-curve-controll-point/39729/14?u=fireball

**Comment by robMate** [comment]
I tested the flatpak 4.4.2, no crash. I build the 4.4.2 local, it does not work. So to me it looks like one of my dependencies/ shared libraries is broken?

**Comment by robMate** [comment]
I learned a bit more gdb and found a solution 🥳 
It looks like the for loop variable gets optimized out.

![Bildschirmfoto vom 2023-10-03 13-24-50](https://github.com/darktable-org/darktable/assets/5657121/fb92c3e8-dc63-451b-8181-2eddf2141a1e)
If i put a ***volatile*** on the k variable in the rgbcurve.c file it works and i can see the variable in gdb.
![Bildschirmfoto vom 2023-10-03 13-15-07](https://github.com/darktable-org/darktable/assets/5657121/93c92006-c6f6-4042-a5bf-2f99506bf833)

 It does not happen on my laptop and by the looks of things to nobody but me 😅 
 
 Is this a real solution, or did i create something wild? I'm not a c developer but it works for now.

**Comment by robMate** [comment]
Ok i narrowed it down to a flag contained in the -O3 optimization group: -fvect-cost-model=dynamic
When i compile it without my fix and with the optimization -O2 plus all optimization from O3 without the -fvect-cost-model=dynamic the rgb curve module is stable. 

Im not sure if something can come from this information, but this is what i found so far.

**Comment by TurboGit** [comment]
So looks like a vectorization issue, @ralfbrown may have an idea?

**Comment by flannelhead** [comment]
At least the module data doesn't seem to be aligned:
https://github.com/darktable-org/darktable/blob/ddc2a5ba2eadbc7b9475a3f760329df0a7604eda/src/iop/rgbcurve.c#L1460
Might be necessary to alloc it aligned? At least it would be a good first guess.

**Comment by ralfbrown** [comment]
`volatile` eliminates most optimizations involving the variable, since it requires that every read in the code emit exactly one read of main memory and every write generate exactly one write to memory.  This doesn't (at first glance) appear to be an alignment issue, so I'd be surprised if aligning the memory allocation will fix it.

At this point, we can't yet rule out a compiler bug, since this seems to happen only when optimizing for the newest architecture.

**Comment by TurboGit** [comment]
Tried on my side with Release mode (so `-O3`) with `GCC 13.2` and no crash. My arch is set to `-march=native` (not sure which one is selected, I'm on x86_64 i9-9980HK).

**Comment by ralfbrown** [comment]
Confirmed fixed by #15742.
